device/esp_tinyusb/descriptors_control.c

```diff
@@ -41,9 +43,58 @@ uint8_t const *tud_descriptor_device_cb(void)
 uint8_t const *tud_descriptor_configuration_cb(uint8_t index)
 {
     (void)index; // Unused, this driver supports only 1 configuration
-    return s_configuration_descriptor;
+    assert(s_desc_cfg);
+    assert(s_desc_cfg->cfg);
+
+#if (TUD_OPT_HIGH_SPEED)
+    // HINT: cfg and hs_cfg are the same, no need to assert(hs_cfg)
+    assert(s_desc_cfg->fs_cfg);
+    // Return configuration descriptor based on Host speed
+    return (TUSB_SPEED_HIGH == tud_speed_get())
+           ? s_desc_cfg->hs_cfg
+           : s_desc_cfg->fs_cfg;
+#else
+    return s_desc_cfg->cfg;
+#endif // TUD_OPT_HIGH_SPEED
 }
 
+#if (TUD_OPT_HIGH_SPEED)
+/**
+ * @brief Invoked when received GET DEVICE QUALIFIER DESCRIPTOR request
+ * Descriptor contents must exist long enough for transfer to complete
+ * If not highspeed capable stall this request
+ */
+uint8_t const *tud_descriptor_device_qualifier_cb(void)
+{
+    assert(s_desc_cfg);
+    assert(s_desc_cfg->qualifier);
+    return (uint8_t const *)s_desc_cfg->qualifier;
+}
+
+
+/**
+ * @brief Invoked when received GET OTHER SPEED CONFIGURATION DESCRIPTOR request
+ * Descriptor contents must exist long enough for transfer to complete
+ * Configuration descriptor in the other speed e.g if high speed then this is for full speed and vice versa
+ */
+uint8_t const *tud_descriptor_other_speed_configuration_cb(uint8_t index)
+{
+    assert(s_desc_cfg);
+    assert(s_desc_cfg->other_speed);
+
+    const uint8_t *other_speed = (TUSB_SPEED_HIGH == tud_speed_get())
+                                 ? s_desc_cfg->fs_cfg
+                                 : s_desc_cfg->hs_cfg;
+
+    memcpy(s_desc_cfg->other_speed,
```

Comment:
Why do we need to memcpy the descriptor? can't we just return to either HS or FS desc?

Reply:
Because we need change the bDescriptorType to 0x07 (OTHER_SPEED_CONFIGURATION) in the buffer, which must exist long enough for transfer to complete.

If we return just the pointer - we return the descriptor with type 0x02 (CONFIGURATION) as a reply for GET_DESCRIPTOR request with descriptor type 0x07  (OTHER_SPEED_CONFIGURATION) in the wValue field.